R/verb-joins.R

```diff
@@ -297,19 +347,108 @@ join_inline_select <- function(lq, by, on) {
   )
 }
 
-update_join_vars <- function(vars, x_lq, y_lq, x_vars, y_vars, type, by) {
-  vars$all_x <- op_vars(x_lq)
-  vars$all_y <- op_vars(y_lq)
+join_needs_new_query <- function(x_lq, join_alias, type) {
+  if (!inherits(x_lq, "lazy_multi_join_query")) {
+    return(TRUE)
+  }
+
+  if (!type %in% c("left", "inner")) {
+    return(TRUE)
+  }
+
+  x_as <- join_alias$x
+  y_as <- join_alias$y
+
+  as_current <- x_lq$table_names$as
+  if (!is_null(x_as)) {
```

Comment:
This feels a bit fiddly to me and I wonder if we're putting too much emphasis on `x_as`/`y_as` which should be relatively rarely used features. I think it would be fine if the last specified alias wins (assuming that's easier to implement)

Reply:
It would still require a check if `sql_on` is used, e.g.

```r
left_join(x, y, x_as = "x", y_as = "y", sql_on = "x.a = y.a") %>% 
  left_join(
    z,
    x_as = "dummy",
    y_as = "z",
    sql_on = "dummy.a = z.a"
  )
```

Reply:
That feels pretty fiddly to me too; I'm not sure we need to support so many edge cases.

Reply:
I agree that it is quite fiddly. But I'm not sure there is a nice alternative. We need some kind of check to make sure that

```r
left_join(lf, lf, sql_on = "...") %>% 
  left_join(lf, sql_on = "...")
```

works. Of course, we could simply start a new query but in the end the checks won't be much simpler. Also, at least for now it feels like more work to change it.

Reply:
Ok, lets leave it as is then.